### ISSUE TITLE
Incorrect variables in dynamic includes in dependent role with allowed duplicates

### ISSUE THREAD

**Issue by sh-vi** [ansible/ansible#23351]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
include module

##### ANSIBLE VERSION
- v2.3.0.0-0.3.rc3:
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]```
```

##### CONFIGURATION
default

##### OS / ENVIRONMENT
Fedora release 25 (Twenty Five)

##### SUMMARY
Dynamic include breaks variables in dependent role. I think it is related with "allow_duplicates: yes" in dependent role. It looks like dynamic include uses the first "scope" of variables.

##### STEPS TO REPRODUCE
- test.yml:
```yaml
---
- hosts: localhost

  roles:
    - role: a
      a:
        name: a1
    - role: a
      a:
        name: a2
```
- roles/a/meta/main.yml:
```yaml
---
dependencies:
  - role: b
    b:
      name: "{{ a.name }}"
```
- roles/a/tasks/main.yml
```yaml
---
- name: output name
  debug:
    msg: "{{ a.name }}"

- include: include.yml
  static: no
```
- roles/a/tasks/include.yml:
```yaml
---
- name: output name in include
  debug:
    msg: "{{ a.name }}"
```
- roles/b/meta/main.yml:
```yaml
---
allow_duplicates: yes
```
- roles/b/tasks/main.yml:
```yaml
---
- name: output name
  debug:
    msg: "{{ b.name }}"

- include: include.yml
  static: no
```
- roles/b/tasks/include.yml:
```yaml
---
- name: output name in include
  debug:
    msg: "{{ b.name }}"
```
I have made test repo: https://github.com/mesk41in/ansible-include-test.

##### EXPECTED RESULTS
In second task "b : output name in include" value of "msg" must be "a2".

##### ACTUAL RESULTS
```
PLAY [localhost] *****************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [b : output name] ***********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a1"
}

TASK [b : include] ***************************************************************************************************************************************************************************************************************************
included: /root/Documents/ansible_include_test/roles/b/tasks/include.yml for localhost

TASK [b : output name in include] ************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a1"
}

TASK [a : output name] ***********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a1"
}

TASK [a : include] ***************************************************************************************************************************************************************************************************************************
included: /root/Documents/ansible_include_test/roles/a/tasks/include.yml for localhost

TASK [a : output name in include] ************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a1"
}

TASK [b : output name] ***********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a2"
}

TASK [b : include] ***************************************************************************************************************************************************************************************************************************
included: /root/Documents/ansible_include_test/roles/b/tasks/include.yml for localhost

TASK [b : output name in include] ************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a1"
}

TASK [a : output name] ***********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a2"
}

TASK [a : include] ***************************************************************************************************************************************************************************************************************************
included: /root/Documents/ansible_include_test/roles/a/tasks/include.yml for localhost

TASK [a : output name in include] ************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "a2"
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=13   changed=0    unreachable=0    failed=0
```

**Comment by SnoopInf** [comment]
Any updates on this? Faced exactly the same issue

**Comment by sh-vi** [comment]
It's look like in ansbile 2.5.2 this bug is already fixed:
```
ansible 2.5.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```
```
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

[DEPRECATION WARNING]: The use of 'static' has been deprecated. Use 'import_tasks' for static inclusion, or 'include_tasks' for dynamic inclusion. This feature will be removed in a future release. Deprecation warnings can be disabled by
 setting deprecation_warnings=False in ansible.cfg.

PLAY [localhost] ****************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [b : output name] **********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a1"
}

TASK [b : include] **************************************************************************************************************************************************************************************************************************
included: /root/Documents/git/q/ansible-include-test/roles/b/tasks/include.yml for localhost

TASK [b : output name in include] ***********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a1"
}

TASK [a : output name] **********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a1"
}

TASK [a : include] **************************************************************************************************************************************************************************************************************************
included: /root/Documents/git/q/ansible-include-test/roles/a/tasks/include.yml for localhost

TASK [a : output name in include] ***********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a1"
}

TASK [b : output name] **********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a2"
}

TASK [b : include] **************************************************************************************************************************************************************************************************************************
included: /root/Documents/git/q/ansible-include-test/roles/b/tasks/include.yml for localhost

TASK [b : output name in include] ***********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a2"
}

TASK [a : output name] **********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a2"
}

TASK [a : include] **************************************************************************************************************************************************************************************************************************
included: /root/Documents/git/q/ansible-include-test/roles/a/tasks/include.yml for localhost

TASK [a : output name in include] ***********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a2"
}

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
localhost                  : ok=13   changed=0    unreachable=0    failed=0
```

**Comment by sh-vi** [comment]
The problem was fixed between v2.4.2.0-1 and v2.4.3.0-1 (backported from v2.5.0a1) by f1c458303a1ad00675497597e1f7fbc0539e314e (b107e397cbd75d8e095b08495da2ac2aad3ef96f). See #34752.
